src/interpreter/runner.ts

```diff
@@ -88,17 +95,38 @@ async function buildDocument(program: string): Promise<BuildResult> {
     }
 }
 
-async function runProgram(program: LoxProgram, outerContext: InterpreterContext): Promise<void> {
+export async function runProgram(program: LoxProgram, outerContext: InterpreterContext): Promise<void> {
+    const cancellationTokenSource = new CancellationTokenSource();
+    const cancellationToken = cancellationTokenSource.token;
+    
+    const timeout = setTimeout(async () => {
+        cancellationTokenSource.cancel();  
+    }, TIMEOUT_MS);
+    
     const context: RunnerContext = {
         variables: new Variables(),
-        log: outerContext.log
+        cancellationToken,
+        timeout,
+        log: outerContext.log,
+        onStart: outerContext.onStart,
     };
-    context.variables.enter();
+
     let end = false;
+    
+    context.variables.enter();
```

Comment:
Minor: Coding style: It is normally a good habit to surround such pair calls like `enter`/`leave` with `try-finally`. Why? Because someone could accidentally write return within the section. Then the finally part is not called. This can cause some debug time, which could be avoided by such a coding habit.
Are there any reasons against it?